llvm/include/llvm/Support/VirtualOutputError.h

```diff
@@ -0,0 +1,132 @@
+//===- VirtualOutputError.h - Errors for output virtualization --*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_SUPPORT_VIRTUALOUTPUTERROR_H
+#define LLVM_SUPPORT_VIRTUALOUTPUTERROR_H
+
+#include "llvm/Support/Error.h"
+#include "llvm/Support/VirtualOutputConfig.h"
+
+namespace llvm {
+namespace vfs {
+
+const std::error_category &output_category();
+
+enum class OutputErrorCode {
+  // Error code 0 is absent. Use std::error_code() instead.
+  not_closed = 1,
+  invalid_config,
+  already_closed,
+};
+
+inline std::error_code make_error_code(OutputErrorCode EV) {
+  return std::error_code(static_cast<int>(EV), output_category());
+}
+
+/// Error related to an \a OutputFile. Derives from \a ECError and adds \a
+/// getOutputPath().
+class OutputError : public ErrorInfo<OutputError, ECError> {
+  void anchor() override;
+
+public:
+  StringRef getOutputPath() const { return OutputPath; }
+  void log(raw_ostream &OS) const override {
+    OS << getOutputPath() << ": ";
+    ECError::log(OS);
+  }
+
+  // Used by ErrorInfo::classID.
+  static char ID;
+
+  OutputError(const Twine &OutputPath, std::error_code EC)
+      : ErrorInfo<OutputError, ECError>(EC), OutputPath(OutputPath.str()) {
+    assert(EC && "Cannot create OutputError from success EC");
+  }
+
+  OutputError(const Twine &OutputPath, OutputErrorCode EV)
+      : ErrorInfo<OutputError, ECError>(make_error_code(EV)),
+        OutputPath(OutputPath.str()) {
+    assert(EC && "Cannot create OutputError from success EC");
+  }
+
+private:
+  std::string OutputPath;
+};
+
+inline Error errorCodeToOutputError(const Twine &OutputPath,
```

Comment:
Nit: I think "createOutputError" is clearer and better matches similar API in Error.h.  Same for `errorCodeToTempFileOutputError` below.

Reply:
Oh, I didn't think the APIs in Error.h ever returned `Error::success()`. Can fix.

Reply:
Actually, I confirmed I was right. `create*Error()` are all actual errors, calling:
```
template <typename ErrT, typename... ArgTs> Error make_error(ArgTs &&... Args) {
  return Error(std::make_unique<ErrT>(std::forward<ArgTs>(Args)...));
}
```
Whereas, my naming is matching `errorCodeToError()`, which can return `Error::success()`.

WDYT?

Reply:
(I didn't add `create*Error()` functions since `make_error()` seemed to do the job!)

Reply:
df630f855359 fills out documentation. Let me know if you're happy "enough with that".

Reply:
> Oh, I didn't think the APIs in Error.h ever returned Error::success(). 

I suppose that's a good argument against "create".

> Whereas, my naming is matching errorCodeToError(), which can return Error::success().

The thing I don't like about this name is that it sounds like a conversion from error_code, but you actually need a path, and the error_code is restricted to certain error codes.  I guess I don't have a better suggestion for now.

Reply:
Pushed f0e274d812e994ed5f201fa53456427f0fe408e6 based on Slack discussion: `convertToOutputError()`.